[Excel] (Custom Functions) Quick Start (#784)
OfficeDev/office-js-docs-pr: docs/quickstarts/excel-custom-functions-quickstart.md

---
ms.date: 03/06/2019
description: Developing custom functions in Excel quick start guide.
title: Custom functions quick start (preview)
localization_priority: Normal
---

> # Get started developing Excel Custom Functions

Reviewer: ```suggestion
# Get started developing Excel custom functions
```

> Custom functions enable developers to add new functions to Excel by defining those functions in JavaScript or Typescript as part of an add-in. Users within Excel can access custom functions just as they would any native function in Excel, such as `SUM()`.

Reviewer: ```suggestion
With custom functions, developers can now add new functions to Excel by defining them in JavaScript or Typescript as part of an add-in. Excel users can access custom functions just as they would any native function in Excel, such as `SUM()`.
```

Reviewer: Per Microsoft Style Writing Guide, "Don't use enables (or allows or lets) to refer to things that Microsoft or Microsoft apps make possible for the customer"

## Prerequisites

[!include[Excel custom functions note](../includes/excel-custom-functions-note.md)]

You'll need the following tools and related resources to begin creating custom functions.

- [Node.js](https://nodejs.org/en/) (version 8.0.0 or later)

- [Git Bash](https://git-scm.com/downloads) (or another Git client)

- The latest version of [Yeoman](https://yeoman.io/) and the [Yeoman generator for Office Add-ins](https://www.npmjs.com/package/generator-office). To install these tools globally, run the following command via the command prompt:

    ```
    npm install -g yo generator-office
    ```

    > [!NOTE]

>     > Even if you have previously installed the Yeoman generator, we recommend updating your package to the latest version from npm.

Reviewer: ```suggestion
    > Even if you've previously installed the Yeoman generator, we recommend you update your package to the latest version from npm.
```

## Build your first custom functions project

To start, you'll use the Yeoman generator to create the custom functions project. This will set up your project with the correct folder structure, source files, and dependencies to begin coding your custom functions.

1. Run the following command and then answer the prompts as follows.

    ```
    yo office
    ```

    - Choose a project type: `Excel Custom Functions Add-in project (...)`

    - Choose a script type: `JavaScript`

    - What do you want to name your add-in? `stock-ticker`

    ![Yeoman generator for Office Add-ins prompts for custom functions](../images/12-10-fork-cf-pic.jpg)

    The Yeoman generator will create the project files and install supporting Node components.

> 2. Go to the project folder.

Reviewer: ```suggestion
2. Navigate to the project folder you just created.
```

    ```
    cd stock-ticker
    ```

> 3. Trust the self-signed certificate that is needed to run this project. For detailed instructions for either Windows or Mac, see [Adding Self Signed Certificates as Trusted Root Certificate](https://github.com/OfficeDev/generator-office/blob/master/src/docs/ssl.md).  

Reviewer: ```suggestion
3. Trust the self-signed certificate you need to run this project. For detailed instructions for either Windows or Mac, see [Adding Self Signed Certificates as Trusted Root Certificate](https://github.com/OfficeDev/generator-office/blob/master/src/docs/ssl.md).  
```

4. Build the project.

    ```
    npm run build
    ```

5. Start the local web server, which runs in Node.js.

>     - If you'll be using Excel for Windows to test your custom functions, run the following command to start the local web server, launch Excel, and sideload the add-in:

Reviewer: ```suggestion
    - If you use Excel for Windows to test your custom functions, run the following command to start the local web server, launch Excel, and sideload the add-in:
```

        ```
         npm run start
        ```
        After running this command, your command prompt will show details about starting the web server. Excel will start with your add-in loaded. If you add-in does not load, check that you have completed step 3 properly.

>     - If you'll be using Excel Online to test your custom functions, run the following command to start the local web server:

Reviewer: ```suggestion
    - If you use Excel Online to test your custom functions, run the following command to start the local web server:
```

        ```
        npm run start-web
        ```

         After running this command, your command prompt will show details about starting the web server. To use your functions, open a new workbook in Excel Online. In this workbook, you'll need to load your add-in. 

>         To do this, select the **Insert** tab on the ribbon and select **Office Add-ins**. Next select **Manage My Add-ins > Upload My Add-in**. Browse for your manifest file and upload it. If your add-in does not load, check you've completed step 3 correctly.

Reviewer: ```suggestion
        To do this, select the **Insert** tab on the ribbon and select **Office Add-ins**. Next, select **Manage My Add-ins > Upload My Add-in**. Browse for your manifest file and upload it. If your add-in does not load, check you've completed step 3 correctly.
```

Reviewer: Are you sure that "Office Add-ins" is on the Insert tab? When I look at mine, it offers two choices "Get Add-ins" and "My Add-ins" but not "Office Add-ins."

Author: ```suggestion
        To do this, select the **Insert** tab on the ribbon and select **Get Add-ins**. In the resulting new window, ensure you are on the **My Add-ins** tab. Next, select **Manage My Add-ins > Upload My Add-in**. Browse for your manifest file and upload it. If your add-in does not load, check you've completed step 3 correctly.
```
I'm still not sure why this is like 5 clicks to just get a thing to sideload, I have always found this very cumbersome to explain. Is that something we should take to the product team and/or UX folk? Good catch btw!

Author: Also - numbers in docs. Do we need to spell out three or is it ok to just put the number?

Reviewer: @mscharlock You can use the numeral for referring to numbered sections in a list. "Step 3" is ok! (See "Numbers" in the Microsoft Style Writing Guide)

## Try out the prebuilt custom functions

The custom functions project that you created by using the Yeoman generator contains some prebuilt custom functions, defined within the **src/customfunctions.js** file. The **manifest.xml** file in the root directory of the project specifies that all custom functions belong to the `CONTOSO` namespace.

> In your Excel workbook, try out the `ADD` custom function by completing the following steps in Excel:

Reviewer: ```suggestion
In your Excel workbook, try out the `ADD` custom function by completing the following steps:
```

> 1. Within a cell, type `=CONTOSO`. Notice that the autocomplete menu shows the list of all functions in the `CONTOSO` namespace.

Reviewer: ```suggestion
1. Select a cell and type `=CONTOSO`. Notice that the autocomplete menu shows the list of all functions in the `CONTOSO` namespace.
```

Author: Random q but why is it that type is ok from an accessibility standards POV? You'd think if we say "select" instead of click everywhere that type similarly suggests having a keyboard structure of some kind (whereas, enter doesn't).

> 2. Run the `CONTOSO.ADD` function, with numbers `10` and `200` as input parameters, by typing the value `=CONTOSO.ADD(10,200)` in the cell and pressing enter.

Reviewer: ```suggestion
2. Run the `CONTOSO.ADD` function, using numbers `10` and `200` as input parameters, by typing the value `=CONTOSO.ADD(10,200)` in the cell and pressing enter.
```

The `ADD` custom function computes the sum of the two numbers that you specify as input parameters. Typing `=CONTOSO.ADD(10,200)` should produce the result **210** in the cell after you press enter.

## Next steps

> Congratulations, you've successfully created a custom function in an Excel add-in! Next, build a more complex add-in with streaming data capability. The below link will take you to the next steps to follow in the Excel add-in with custom functions tutorial.

Reviewer: ```suggestion
Congratulations, you've successfully created a custom function in an Excel add-in! Next, build a more complex add-in with streaming data capability. The following link takes you through the next steps in the Excel add-in with custom functions tutorial.
```

Author: Oh man a below got through my internal edit! Dangnabit!

---
ms.date: 01/08/2019
description: Learn best practices for developing custom functions in Excel.
title: Custom functions best practices (preview)
localization_priority: Normal
---

# Get started developing Excel Custom Functions

Custom functions enable developers to add new functions to Excel by defining those functions in JavaScript or Typescript as part of an add-in. Users within Excel can access custom functions just as they would any native function in Excel, such as `SUM()`.

> ## Platforms

Author: Oh right! Good idea

    ```

    - Choose a project type: `Excel Custom Functions Add-in project (...)`

    - Choose a script type: `JavaScript`

    - What do you want to name your add-in? `stock-ticker`

    ![Yeoman generator for Office Add-ins prompts for custom functions](../images/12-10-fork-cf-pic.jpg)

    The Yeoman generator will create the project files and install supporting Node components. The project files come from the [Excel-Custom-Functions](https://github.com/OfficeDev/Excel-Custom-Functions) GitHub repository.

2. Go to the project folder.

    ```
    cd stock-ticker
    ```

3. Trust the self-signed certificate that is needed to run this project. For detailed instructions for either Windows or Mac, see [Adding Self Signed Certificates as Trusted Root Certificate](https://github.com/OfficeDev/generator-office/blob/master/src/docs/ssl.md).  

4. Build the project.

    ```
    npm run build
    ```

5. Start the local web server, which runs in Node.js.

    - If you'll be using Excel for Windows to test your custom functions, run the following command to start the local web server, launch Excel, and sideload the add-in:

        ```
         npm run start
        ```

>         After running this command, your command prompt will show details about what has been done, another npm window will open showing the details of the build, and Excel will start with your add-in loaded. If you add-in does not load, check that you have completed step 3 properly.

Reviewer 2: ```suggestion
        After running this command, your command prompt will show details about starting the web server. Excel will start with your add-in loaded. If you add-in does not load, check that you have completed step 3 properly.
```

I'm not seeing a second window launch so just updating this.

    - If you'll be using Excel Online to test your custom functions, run the following command to start the local web server:

        ```
        npm run start-web
        ```

>          After running this command, another window will open showing you the details of the build. To use your functions, open a new workbook in Office Online.

Reviewer 2: Darn, for online you will have to sideload. We should copy in those steps here.

## Try out the prebuilt custom functions

The custom functions project that you created by using the Yeoman generator contains some prebuilt custom functions, defined within the **src/customfunctions.js** file. The **manifest.xml** file in the root directory of the project specifies that all custom functions belong to the `CONTOSO` namespace.

In your Excel workbook, try out the `ADD` custom function by completing the following steps in Excel:

1. Within a cell, type `=CONTOSO`. Notice that the autocomplete menu shows the list of all functions in the `CONTOSO` namespace.

2. Run the `CONTOSO.ADD` function, with numbers `10` and `200` as input parameters, by typing the value `=CONTOSO.ADD(10,200)` in the cell and pressing enter.

The `ADD` custom function computes the sum of the two numbers that you specify as input parameters. Typing `=CONTOSO.ADD(10,200)` should produce the result **210** in the cell after you press enter.

## Next steps

Congratulations, you've successfully created a custom function in an Excel add-in! Next, learn more about the capabilities of custom functions and build a more complex add-in by following along with the Excel custom functions add-in tutorial.

> 

Reviewer 2: In the tutorial we should add a note that if you already did the quick start you can jump to the next section since it repeats a lot of steps here.

Author: I went ahead and made a link that jumps right to the right section for people - or should I change it back and have a note?
